### DISCUSSION TITLE
Limit refspecs to "recent" pull requests

### DISCUSSION THREAD

**Discussion by DamienCassou** [magit/forge#627] [Ideas]
I'm working on a project with more than ten thousand PRs: 25 are currently open and they are within the most recent 500 ones. Listing branches in magit is quite slow because of all the `refs/pullreqs` refspecs.

To make this faster, I started filtering the refspecs that are fetched:

```git-config
[remote "origin"]
	fetch = +refs/heads/*:refs/remotes/origin/*
	fetch = +refs/pull/*/head:refs/pullreqs/*
	fetch = ^refs/pull/2*/head # don't fetch old PRs
	fetch = ^refs/pull/3*/head # don't fetch old PRs
	fetch = ^refs/pull/4*/head # don't fetch old PRs
	fetch = ^refs/pull/5*/head # don't fetch old PRs
	fetch = ^refs/pull/6*/head # don't fetch old PRs
	fetch = ^refs/pull/7*/head # don't fetch old PRs
	fetch = ^refs/pull/8*/head # don't fetch old PRs
	fetch = ^refs/pull/9*/head # don't fetch old PRs
```

This filters out all PRs that don't start with a "1". This works because all recent PRs (those greater than `#12131`) start with a "1".

Then I cleaned all PR refspecs with:

```session
$ git for-each-ref --format='delete %(refname)' refs/pullreqs | git update-ref --stdin
```

Then I launched `forge-pull` again to fetch again all PR refspecs starting with a "1".

This makes magit much faster but is very hackish. Do you have any idea on how to do better?

**Comment by tarsius** [comment]
> Listing branches in magit is quite slow because of all the `refs/pullreqs` refspecs.

What exactly gets slower? `magit-show-refs`?

`git for-each-ref "--format=... refs/heads` (using the same format as `magit-insert-local-branches`) takes 0.3 seconds instead of 0.12 for https://github.com/NixOS/nixpkgs. Is that the slowdown you are experiencing, or do you see a bigger effect? (I haven't imported any api data into forge.)

`magit-status` takes quite a while. Unfortunately I forgot to time before adding pull-request refspec, so I am not cloning a second time. (Which takes a while.)

**Reply by DamienCassou** [reply]
`magit-checkout` for example was very slow to list branches to switch to.

But, following your question, I have done more tests and I have no clue what is happening anymore. For example, on the nixpkgs repository, `magit-checkout` takes between 1 second and around 7 seconds to list only 661 refs. Why these differences in speed? 

It seems that one slow thing is `magit-get-previous-branch`:

```emacs
(benchmark-call
  (lambda ()
    (let ((default-directory "/home/cassou/Documents/projects/nix/nixpkgs-master"))
      (magit-get-previous-branch)))
  20.0)
```

This results in `(21 36.658059883 0 0.0)` showing that each call to `magit-get-previous-branch` takes around 1s.

**Reply by tarsius** [reply]
I'll look into that function, but it will have to wait until I am less busy.

**Reply by tarsius** [reply]
> It seems that one slow thing is `magit-get-previous-branch`

This function could take a long time, so we added a timeout in 2018. See https://github.com/magit/magit/issues/3457. Does this not work for you, or have you customized `magit-get-previous-branch-timeout`? Try adding a debug statement to measure how long a single call to `magit-rev-branch` takes, and how many revisions are tried.

**Reply by tarsius** [reply]
Also:

To be able to focus on other projects, I have to reduce the time I spend on Forge. That involves deciding to not implement certain features and I'm afraid "fetch only some of the pr refs" is one of those. Sorry.

If there is an issue with `magit-get-previous-branch-timeout`, I would still be interested in fixing that.

**Reply by DamienCassou** [reply]
Understood. I will open an issue if I find one.
